crates/ra_assists/src/handlers/fill_match_arms.rs

```diff
@@ -56,31 +50,115 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
+                    let pat = arm.pat().unwrap();
+
+                    // Special casee OrPat as separate top-level pats
+                    let pats: Vec<Pat> = match Pat::from(pat.clone()) {
+                        Pat::OrPat(pats) => pats.pats().collect::<Vec<_>>(),
+                        _ => vec![pat],
+                    };
+
+                    pats.iter().any(|pat| {
+                        match does_arm_pat_match_variant(pat, arm.guard(), variant_pat) {
+                            ArmMatch::Yes => true,
+                            ArmMatch::No => false,
+                            ArmMatch::Partial => {
+                                has_partial_match = true;
+                                true
+                            }
+                        }
+                    })
+                })
+            })
+            .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()))
+            .collect();
+
+        arms.extend(variants);
+        if has_partial_match {
+            arms.push(make::match_arm(
+                iter::once(make::placeholder_pat().into()),
+                make::expr_unit(),
+            ));
+        }
+
+        let indent_level = IndentLevel::from_node(match_arm_list.syntax());
+        let new_arm_list = indent_level.increase_indent(make::match_arm_list(arms));
 
         edit.target(match_expr.syntax().text_range());
         edit.set_cursor(expr.syntax().text_range().start());
         edit.replace_ast(match_arm_list, new_arm_list);
     })
 }
 
-fn is_trivial(arm: &ast::MatchArm) -> bool {
-    match arm.pat() {
-        Some(ast::Pat::PlaceholderPat(..)) => true,
-        _ => false,
+enum ArmMatch {
+    Yes,
+    No,
+    Partial,
+}
+
+fn does_arm_pat_match_variant(arm: &Pat, arm_guard: Option<MatchGuard>, var: &Pat) -> ArmMatch {
+    let arm = flatten_pats(arm.clone());
+    let var = flatten_pats(var.clone());
+    let mut arm = arm.iter();
+    let mut var = var.iter();
+
+    // If the first part of the Pat don't match, there's no match
+    match (arm.next(), var.next()) {
+        (Some(arm), Some(var)) if arm.text() == var.text() => {}
+        _ => return ArmMatch::No,
+    }
+
+    // If we have a guard we automatically know we have a partial match
+    if arm_guard.is_some() {
+        return ArmMatch::Partial;
+    }
+
+    if arm.clone().count() != var.clone().count() {
+        return ArmMatch::Partial;
+    }
+
+    let direct_match = arm.zip(var).all(|(arm, var)| {
+        if arm.text() == var.text() {
+            return true;
+        }
+        match (arm.kind(), var.kind()) {
+            (SyntaxKind::PLACEHOLDER_PAT, SyntaxKind::PLACEHOLDER_PAT) => true,
+            (SyntaxKind::DOT_DOT_PAT, SyntaxKind::PLACEHOLDER_PAT) => true,
+            (SyntaxKind::BIND_PAT, SyntaxKind::PLACEHOLDER_PAT) => true,
+            _ => false,
+        }
+    });
+
+    match direct_match {
```

Comment:
Funny, an asisst does not apply here.
Also it feels like a plan `if` would suffice, but up to you.

Reply:
Same reason as above. I could potentially implement a subset of other types, let me know if you have any suggestions...

Are plain `if`s more idiomatic for cases like this?

Reply:
Right, missed that reason totally, thanks.

I think `idiomatic` may be very subjective sometimes, for me it looks a bit of an overkill to match over the `bool` variable. But this is a minor thing, up to you to change or not.

My biggest concerns about this PR are the unreachable branch added and the method refactoring comment, those would be cool to fix.